### PR TITLE
Updates div to perform true division in pt/xla

### PR DESCRIPTION
**Pull request by JackCaoG**
This is to resolve the https://github.com/pytorch/xla/issues/2452. The caveat right now is non-float inplace div would result in a runtime error.

### CODE REVIEW

File: torch_xla/csrc/aten_xla_type.cpp [hunk 1]
```diff
@@ -995,20 +995,16 @@ at::Tensor AtenXlaType::diagonal(const at::Tensor& self, int64_t offset,
 
 at::Tensor AtenXlaType::div(const at::Tensor& self, const at::Tensor& other) {
   XLA_FN_COUNTER("xla::");
-  return DoBinaryOp(self, other,
-                    [&](const XLATensor& xself, const XLATensor& xother,
-                        at::ScalarType dtype) {
-                      return XLATensor::div(xself, xother, dtype);
-                    });
+  at::ScalarType dtype = at::result_type(self, other);
+  std::pair<XLATensor, XLATensor> operands = GetBinaryOperands(self, other);
```
Comment (davidel) [inline torch_xla/csrc/aten_xla_type.cpp:999]: We use `auto` usually for templated locals.

File: torch_xla/csrc/tensor_methods.cpp [hunk 2]
```diff
@@ -957,25 +969,54 @@ XLATensor XLATensor::diagonal(const XLATensor& input, xla::int64 offset,
...
+  xla::PrimitiveType input_type = input.GetIrValue().shape().element_type();
+  xla::PrimitiveType other_type = other.GetIrValue().shape().element_type();
+  bool input_is_float = xla::primitive_util::IsFloatingPointType(input_type);
+  bool other_is_float = xla::primitive_util::IsFloatingPointType(other_type);
+  if (input_is_float && !other_is_float) {
+    scalar_type = TensorTypeFromXlaType(input_type);
+  } else if (!input_is_float && other_is_float) {
+    scalar_type = TensorTypeFromXlaType(other_type);
+  }
+  ir::Value input_value = GetFloatingIrValue(input, scalar_type);
```
Comment (davidel) [inline torch_xla/csrc/tensor_methods.cpp:983]: Make the `GetFloatingIrValue()` API to return an `std::pair<ir::Value, ir::Value>`, returning the original and possibly casted values.
Then get the original type from the original value shape.

**Review by davidel — CHANGES_REQUESTED** [review]

File: torch_xla/csrc/tensor_methods.cpp (hunk 2 above)
Comment (JackCaoG) [reply torch_xla/csrc/tensor_methods.cpp:1013]: I think changing the `GetFloatingIrValue()` api will cause additional expense to the `div(const XLATensor& input, at::Scalar other` and `div_` as they don't need to know the original type. 

What's the advantage of updating the api to return the pair? I think we need to know the `scalar_type` before calling `GetFloatingIrValue` and we don't really care about the original type after that right?

Comment (davidel) [reply torch_xla/csrc/tensor_methods.cpp:1013]: Hmm, either the code changed or I mis-read it 😄 
No need to change that API.

File: torch_xla/csrc/tensor_methods.cpp [hunk 3]
```diff
@@ -987,25 +999,54 @@ XLATensor XLATensor::diagonal(const XLATensor& input, xla::int64 offset,
 
 XLATensor XLATensor::div(const XLATensor& input, const XLATensor& other,
                          c10::optional<at::ScalarType> logical_element_type) {
-  return input.CreateFrom(input.GetIrValue() / other.GetIrValue(),
-                          logical_element_type);
+  at::ScalarType scalar_type =
+      at::typeMetaToScalarType(c10::get_default_dtype());
+  xla::PrimitiveType input_type = input.GetIrValue().shape().element_type();
```
Comment (davidel) [inline torch_xla/csrc/tensor_methods.cpp:1004]:
```c++
input.shape().get().element_type()
```
Same below.

**Review by davidel — APPROVED** [review]
Couple of minor nits to fix, but pre-approving to bunny-hop the timezone 😄

**Comment by JackCaoG** [comment]
@mruberry I think the xla side is ready

**Review by davidel — APPROVED** [review]

**Pull request merged**